torch/nn/functional.py

```diff
@@ -2316,6 +2316,63 @@ def poisson_nll_loss(input, target, log_input=True, full=False, size_average=Non
     return ret
 
 
+def gaussian_nll_loss(input, target, var, eps=1e-8, full=True, reduction='mean'):
+    r"""Gaussian negative log likelihood loss.
+
+    See :class:`~torch.nn.GaussianNLLLoss` for details.
+
+    Args:
+        input: expectation of the Gaussian distribution.
+        target: sample from the Gaussian distribution.
+        var: tensor of positive variance(s), one for each of the expectations
+        in the input (heteroscedastic), or a single one (homoscedastic).
+        eps: value added to var, for stability.
+        full: ``True``/``False`` (bool), include the constant term in the loss
+            calculation. Default: ``False``.
+        reduction: specifies the reduction to apply to the output:
+            ``'none'`` | ``'mean'`` | ``'sum'``. ``'none'``: no reduction will be applied,
+            ``'mean'``: the output is the average of all batch member losses,
+            ``'sum'``: the output is the sum of all batch member losses.
+            Default: ``'mean'``.
+    """
+    # Inputs and targets much have same shape
+    input = input.view(input.size(0), -1)
+    target = target.view(target.size(0), -1)
+    if input.size() != target.size():
+        raise ValueError("input and target must have same size")
+
+    # Second dim must be 1 ie shape (N, 1) if var and input do not have same shape
+    var = var.view(var.size(0), -1)
+    if var.size() != input.size() and var.size(1) != 1:
+        raise ValueError("var is of incorrect size")
+
+    if reduction != 'none' and reduction != 'mean' and reduction != 'sum':
+        raise ValueError(reduction + " is not valid")
+
+    # Entries of var must be non-negative
+    if torch.any(var < 0):
+        raise ValueError("var has negative entry/entries")
+
+    # Add eps for stability
+    var = var + eps
```

Comment:
nit: Isn't it better to use clamp here instead of add? That way you won't modify values that don't need to be modified.

You can even do that in `torch.no_grad()` if you think that the clamping should be ignored wrt autograd.

Reply:
`var` is the output of an NN here so requires grad, so I can't use clamp as it'll kill the gradients at the clamped values.

Reply:
That is why I mentioned above that you don't have to do that in a differentiable way and you can do:
```python
# Avoid changing the argument given by the user inplace
var = var.clone()
with torch.no_grad():
  var.clamp_(min=eps)
```